video/out/vulkan/context_display.c

```diff
@@ -278,14 +285,88 @@ struct priv {
     struct vulkan_display_opts *opts;
     uint32_t width;
     uint32_t height;
+
+#if HAVE_DRM
+    struct mpv_opengl_drm_params_v2 drm_params;
+#endif
 };
 
+#if HAVE_DRM
+static bool open_render_fd(struct ra_ctx *ctx, int kms_fd)
+{
+    struct priv *p = ctx->priv;
+    p->drm_params.fd = -1;
+    p->drm_params.render_fd = -1;
+
+    char *render_path = drmGetRenderDeviceNameFromFd(kms_fd);
+    int fd = open(render_path, O_RDWR | O_CLOEXEC);
+    free(render_path);
+
+    p->drm_params.render_fd = fd;
+    return fd != -1;
+}
+
+static bool drm_setup(struct ra_ctx *ctx, int display_idx,
+                      VkPhysicalDevicePCIBusInfoPropertiesEXT *pci_props)
+{
+    int fd, card_no = -1;
+    do {
+        int ret;
+        char card_path[128];
+        card_no++;
+        snprintf(card_path, sizeof(card_path),
+                 DRM_DEV_NAME, DRM_DIR_NAME, card_no);
+        fd = open(card_path, O_RDWR | O_CLOEXEC);
+        MP_DBG(ctx, "Opening card: %s | %d\n", card_path, fd);
+        if (fd >= 0) {
+            drmDevicePtr device;
+            ret = drmGetDevice(fd, &device);
+            if (ret == 0 &&
+                device->businfo.pci->domain == pci_props->pciDomain &&
+                device->businfo.pci->bus == pci_props->pciBus &&
+                device->businfo.pci->dev == pci_props->pciDevice &&
+                device->businfo.pci->func == pci_props->pciFunction)
+            {
+                // Found our matching device.
+                MP_DBG(ctx, "Card is matched at %04X:%02X:%02X:%02X\n",
+                       pci_props->pciDomain, pci_props->pciBus,
+                       pci_props->pciDevice, pci_props->pciFunction);
+                drmFreeDevice(&device);
+                break;
+            }
+            drmFreeDevice(&device);
+            close(fd);
+        }
+    } while (fd >= 0);
```

Comment:
Instead of walking the primary nodes to get the device into, to then get the render nodes just use drmGetDevices2 from the start. The following pseudo code example, also checks the bus type that the original code was missing.

```
drmDevice *devs[foo] = {};
int count = drmGetDevices2(0, devs, array_size(devs));

// error out on <= 0

for (i = 0; i < count; i++) {
  drmDevice *dev = devs[i];
  if (!(dev->avail_nodes & RENDER))
    continue;

  fd = open(dev->nodes[RENDER], ...)
  if (fd < 0)
    continue;

  if (dev->bus_type != DRM_BUS_PCI)
    continue;
  // match the businfo ...
  if matched
    p->drm_params.render_fd = fd;
}
drmFreeDevices(...);
```

Reply:
Thanks for pointing this out! I've updated the code.